src/autogen/agen5/autogen.1
.TH autogen 1 "09 Jan 2014" "GNU AutoGen (5.15)" "User Commands"
.\"
.\"  DO NOT EDIT THIS FILE   (opts.man)
.\"  
.\"  It has been AutoGen-ed  January  9, 2014 at 09:33:00 PM by AutoGen 5.15
.\"  From the definitions    ./opts.def
.\"  and the template file   agman-cmd
.\"
.SH NAME
autogen \- The Automated Program Generator
.SH SYNOPSIS
.B autogen
.\" Mixture of short (flag) options and long options
.RB [ \-\fIflag\fP " [\fIvalue\fP]]... [" \-\-\fIopt\-name\fP " [[=| ]\fIvalue\fP]]..." [ <def-file> ]
.PP
AutoGen creates text files from templates using external definitions.
.SH DESCRIPTION
\fBAutoGen\fP is designed for generating program files that contain
repetitive text with varied substitutions.  The goal is to simplify the
maintenance of programs that contain large amounts of repetitious text.
This is especially valuable if there are several blocks of such text
that must be kept synchronized.
.sp
One common example is the problem of maintaining the code required for
processing program options.  Processing options requires a minimum of
four different constructs be kept in proper order in different places
in your program.  You need at least: The flag character in the flag
string, code to process the flag when it is encountered, a global
state variable or two, and a line in the usage text.
You will need more things besides this if you choose to implement
long option names, configuration file processing, environment variables
and so on.
.sp
All of this can be done mechanically; with the proper templates
and this program.
.SH "OPTIONS"
.TP
.BR \-L " \fIdir\fP, " \-\-templ\-dirs "=" \fIdir\fP
Template search directory list.
This option may appear an unlimited number of times.
.sp
Add a directory to the list of directories to search when opening
a template, either as the primary template or an included one.
The last entry has the highest priority in the search list.
That is to say, they are searched in reverse order.
.TP
.BR \-T " \fItpl\-file\fP, " \-\-override\-tpl "=" \fItpl\-file\fP
Override template file.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
Definition files specify the standard template that is to be expanded.
This option will override that name and expand a different template.
.TP
.BR \-l " \fItpl\-file\fP, " \-\-lib\-template "=" \fItpl\-file\fP
Library template file.
This option may appear an unlimited number of times.
.sp
DEFINE macros are saved from this template file for use in processing
the main macro file.  Template text aside from the DEFINE macros is
is ignored.
.TP
.BR \-b " \fIname\fP, " \-\-base\-name "=" \fIname\fP
Base name for output file(s).
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
A template may specify the exact name of the output file.  Normally,
it does not.  Instead, the name is composed of the base name of the
definitions file with suffixes appended.  This option will override the
base name derived from the definitions file name.  This is required if
there is no definitions file and advisable if definitions are being
read from stdin.  If the definitions are being read from standard in,
the base name defaults to \fIstdin\fP.  Any leading directory components
in the name will be silently removed.  If you wish the output file to
appear in a particular directory, it is recommended that you "cd" into
that directory first, or use directory names in the format specification
for the output suffix lists, see: pseudo macro.
.TP
.BR \-\-definitions "=\fIfile\fP", " \fB\-\-no\-definitions\fP"
Definitions input file.
The \fIno\-definitions\fP form will disable the option.
This option is enabled by default.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
Use this argument to specify the input definitions file with a
command line option.  If you do not specify this option, then
there must be a command line argument that specifies the file,
even if only to specify stdin with a hyphen (\fB\-\fP).
Specify, \fB\-\-no\-definitions\fP when you wish to process
a template without any active AutoGen definitions.\\n
.TP
.BR \-S " \fIfile\fP, " \-\-load\-scheme "=" \fIfile\fP
Scheme code file to load.
.sp
Use this option to pre\-load Scheme scripts into the Guile
interpreter before template processing begins.
Please note that the AutoGen specific functions are not loaded
until after argument processing.  So, though they may be specified
in lambda functions you define, they may not be invoked until after
option processing is complete.
.TP
.BR \-F " \fIfile\fP, " \-\-load\-functions "=" \fIfile\fP
Load scheme function library.
.sp
This option is used to load Guile\-scheme functions.  The automatically
called initialization routine \fBscm_init\fP must be used to register
these routines or data.
.TP
.BR \-s " \fIsuffix\fP, " \-\-skip\-suffix "=" \fIsuffix\fP
Omit the file with this suffix.
This option may appear an unlimited number of times.
This option may not be preset with environment variables
or in initialization (rc) files.
This option must not appear in combination with any of the following options:
select\-suffix.
.sp
Occasionally, it may not be desirable to produce all of the output
files specified in the template.  (For example, only the \fI.h\fP
header file, but not the \fI.c\fP program text.)  To do this
specify \fB\-\-skip\-suffix=c\fP on the command line.
.TP
.BR \-o " \fIsuffix\fP, " \-\-select\-suffix [ =\fIsuffix\fP ]
specify this output suffix.
This option may appear an unlimited number of times.
This option may not be preset with environment variables
or in initialization (rc) files.
.sp
If you wish to override the suffix specifications in the template,
you can use one or more copies of this option.  See the suffix
specification in the @ref{pseudo macro} section of the info doc.
.TP
.BR \-\-source\-time, " \fB\-\-no\-source\-time\fP"
set mod times to latest source.
The \fIno\-source\-time\fP form will disable the option.
.sp
If you stamp your output files with the \fBDNE\fP macro output, then
your output files will always be different, even if the content has
not really changed.  If you use this option, then the modification
time of the output files will change only if the input files change.
This will help reduce unneeded builds.
.TP
.BR \-m ", " \-\-no\-fmemopen
Do not use in\-mem streams.
.sp
If the local C library supports "\fBfopencookie(3GNU)\fP", or
"\fBfunopen(3BSD)\fP" then AutoGen prefers to use in\-memory stream
buffer opens instead of anonymous files.  This may lead to problems
if there is a shortage of virtual memory.  If, for a particular
application, you run out of memory, then specify this option.
This is unlikely in a modern 64\-bit virtual memory environment.
.sp
On platforms without these functions, the option is accepted
but ignored.  \fBfmemopen(POSIX)\fP is not adequate because
its string buffer is not reallocatable.  \fBopen_memstream(POSIX)\fP
is \fIalso\fP not adequate because the stream is only opened for
output.  AutoGen needs a reallocatable buffer available for both
reading and writing.
.TP
.BR \-\-equate "=\fIchar\-list\fP"
characters considered equivalent.
The default \fIchar\-list\fP for this option is:
.ti +4
 _\-^
.sp
This option will alter the list of characters considered equivalent.
The default are the three characters, "_\-^".  (The last is conventional
on a Tandem/HP\-NonStop, and I used to do a lot of work on Tandems.)
.TP
.BR \-\-writable, " \fB\-\-not\-writable\fP"
Allow output files to be writable.
The \fInot\-writable\fP form will disable the option.
.sp
This option will leave output files writable.
Normally, output files are read\-only.
.PP
The following options are often useful while debugging new templates:
.TP
.BR \-\-loop\-limit "=\fIlim\fP"
Limit on increment loops.
This option takes an integer number as its argument.
The value of \fIlim\fP is constrained to being:
.in +4
.nf
.na
exactly \-1, or
in the range  1 through 0x1000000
.fi
.in -4
The default \fIlim\fP for this option is:
.ti +4
 256
.sp
This option prevents runaway loops.  For example, if you accidentally
specify, "FOR x (for\-from 1) (for\-to \-1) (for\-by 1)", it will take a
long time to finish.  If you do have more than 256 entries in tables,
you will need to specify a new limit with this option.
.TP
.BR \-\-shell "=\fIshell\fP"
name or path name of shell to use.
.sp
By default, when AutoGen is built, the configuration is probed for a
reasonable Bourne\-like shell to use for shell script processing.  If
a particular template needs an alternate shell, it must be specified
with this option on the command line, with an environment variable
(\fBSHELL\fP) or in the configuration/initialization file.
.TP
.BR \-t " \fItime\-lim\fP, " \-\-timeout "=" \fItime\-lim\fP
Time limit for server shell.
This option takes an integer number as its argument.
The value of \fItime\-lim\fP is constrained to being:
.in +4
.nf
.na
in the range  0 through 3600
.fi
.in -4
.sp
AutoGen works with a shell server process.  Most normal commands will
complete in less than 10 seconds.  If, however, your commands need more
time than this, use this option.
.sp
The valid range is 0 to 3600 seconds (1 hour).
Zero will disable the server time limit.
.TP
.BR \-\-trace "=\fIlevel\fP"
tracing level of detail.
This option takes a keyword as its argument.  The argument sets an enumeration value that can
be tested by comparing them against the option value macro.
The available keywords are:
.in +4
.nf
.na
nothing       debug\-message server\-shell
templates     block\-macros  expressions
everything
.fi
or their numeric equivalent.
.in -4
.sp
The default \fIlevel\fP for this option is:
.ti +4
 nothing
.sp
This option will cause AutoGen to display a trace of its template
processing.  There are six levels, each level including messages from
the previous levels:
.sp
.in +4
.ti -4
.IR nothing
Does no tracing at all (default)
.sp
.ti -4
.IR debug\-message
Print messages from the "DEBUG" AutoGen macro (see: DEBUG).
.sp
.ti -4
.IR server\-shell
Traces all input and output to the server shell.  This includes a shell
"independent" initialization script about 30 lines long.  Its output is
discarded and not inserted into any template.
.sp
.ti -4
.IR templates
Traces the invocation of \fBDEFINE\fPd macros and \fBINCLUDE\fPs
.sp
.ti -4
.IR block\-macros
Traces all block macros.  The above, plus \fBIF\fP, \fBFOR\fP,
\fBCASE\fP and \fBWHILE\fP.
.sp
.ti -4
.IR expressions
Displays the results of expression evaluations.
.sp
.ti -4
.IR everything
Displays the invocation of every AutoGen macro, even \fBTEXT\fP macros
(i.e. the text outside of macro quotes).  Additionally, if you rebuild
the ``expr.ini'' file with debugging enabled, then all calls to
AutoGen defined scheme functions will also get logged:
.br
.br
.in +4
.nf
cd ${top_builddir}/agen5
DEBUG_ENABLED=true bash bootstrap.dir expr.ini
make CFLAGS='\-g \-DDEBUG_ENABLED=1'
.in -4
.fi
.sp
Be aware that you cannot rebuild this source in this way without first
having installed the \fBautogen\fP executable in your search path.
Because of this, "expr.ini" is in the distributed source list, and
not in the dependencies.
.in -4
.TP
.BR \-\-trace\-out "=\fIfile\fP"
tracing output file or filter.
.sp

ChangeLog
2014-01-10  Eric Gallager  <egall@gwmail.gwu.edu>

	add some more old files from the 1990s

	also:
	- make some more makefiles respect @datarootdir@
	- update changelog
	- minor configure tweaks
	- fix some arch-specific stuff for gdbserver
	- silence some warnings in nindy-share now that I have a
	(somewhat old, but working) Makefile for that subdir
	- use autoheader for src/gdb/nlm
	- fix some makefiles that would ignore archflags
	- readline: properly clean up after a shlib build, and stop using
	the `-v` flag when compiling (it just prints the version of gcc,
	which makes output harder to read)
	- src/sim: avoid using -Wuninitialized without optimization
	- src/sim/testsuite: actually use maintainer mode
	- src/tcl/unix/dltest is messed up, so mess it up even more
	(Now TWO separate configure scripts try to AM_INIT_AUTOMAKE its
	Makefile.am!)

2014-01-09  Eric Gallager  <egall@gwmail.gwu.edu>

	minor gitignore update

	in src/autogen/add-on

2014-01-09  Eric Gallager  <egall@gwmail.gwu.edu>

	autogen: add stuff from its sourceforge git

	also in autogen, switch to gnits strictness, and stub in dejagnu
	tests

2014-01-09  Eric Gallager  <egall@gwmail.gwu.edu>

	actually vendor in autogen now

	autogen is needed to regenerate src/Makefile.in and
	src/utils/Makefile.in (apparently they are too complicated to just
	use automake), and I failed to get it (autogen) installed
	separately...

2014-01-09  Eric Gallager  <egall@gwmail.gwu.edu>

	regenerate things with autogen (the guile-based one)

	this is in preparation for vendoring in a copy of autogen for this
	purpose

2014-01-09  Eric Gallager  <egall@gwmail.gwu.edu>

	should have been part of previous commit

	and gitignored, anyways...(?)

2014-01-09  Eric Gallager  <egall@gwmail.gwu.edu>

	found some stuff

	- switch more automake-using subdirs to gnits strictness
	- found sources for amd-udi and msdos in utils
	- found some old doc-generating scripts for bfd
	- other misc tweaks

2014-01-07  Eric Gallager  <egall@gwmail.gwu.edu>

	misc. build fixes

	- Make Makefile.in files respect datarootdir
	- Make configure scripts look in our specific intl subdir
	- skip building zlib from src directory
	- various .c files need to include <libintl.h> directly, apparently
	- move to gnits strictness in intl
	- gnulib was missing some files due to .gitignore files messing
	stuff up

2014-01-06  Eric Gallager  <egall@gwmail.gwu.edu>

	renaming

2014-01-06  Eric Gallager  <egall@gwmail.gwu.edu>

	latest texinfo.tex macro file

2014-01-06  Eric Gallager  <egall@gwmail.gwu.edu>

	add some misc texinfo things

2014-01-06  Eric Gallager  <egall@gwmail.gwu.edu>

	zlib tweaks

2014-01-05  Eric Gallager  <egall@gwmail.gwu.edu>

	dejagnu tweaks

2014-01-05  Eric Gallager  <egall@gwmail.gwu.edu>

	fix building tcl with xcode project

2014-01-01  Eric Gallager  <egall@gwmail.gwu.edu>

	work on sim and Tcl

2014-01-01  Eric Gallager  <egall@gwmail.gwu.edu>

	do some clang static analysis

	also fix readline

2013-12-30  Eric Gallager  <egall@gwmail.gwu.edu>

	I thought I had already made this change previously...

2013-12-30  Eric Gallager  <egall@gwmail.gwu.edu>

	deinit submodules (to save space temporarily)

2013-12-19  Eric Gallager  <egallager@comcast.net>

	touch/un-obsolete old m68k config files

	Change-Id: Iaf1c2456aef5a91ccc8983133f8b4b22fe2cbcbe

2013-12-19  Eric Gallager  <egallager@comcast.net>

	touch more old mips config files

	Change-Id: Iff991ce2549e61749c945c1c25fad91b7674b939

2013-12-18  Eric Gallager  <egallager@comcast.net>

	touch old mips config files

	Change-Id: If238e03939f14bdd0603a2728c59b3a4c9bcdf95

2013-12-18  Eric Gallager  <egallager@comcast.net>

	touch/unobsolete old PPC config files

	Change-Id: If8cdc5a81d87f5d41eb2d1f8d6f597f3df0ef421

2013-12-18  Eric Gallager  <egallager@comcast.net>

	touch old sparc config files

	Change-Id: I01b3911377e6ee95c07dada8ac19b876d0965267

2013-12-17  Eric Gallager  <egallager@comcast.net>

	fix some warnings in utils/sparclite and utils/wince

	Change-Id: I072a258a9b1b3ff5862ccee7653aa2c34f1e5db0

2013-12-17  Eric Gallager  <egallager@comcast.net>

	convert amd-udi to automake

	Change-Id: I0ae2332f009099105dd271a4d12bfb196e11f873

2013-12-17  Eric Gallager  <egallager@comcast.net>

	Try to figure out what is going on in utils/amd-udi

	Change-Id: I7c2dda10ec5229a5b9c4c7951fcb41ead748a404

2013-12-17  Eric Gallager  <egallager@comcast.net>

	more fighting with autoconf's cache mechanism

	Change-Id: I0450706d8e4f2e3d296192635de9776b73c454b4

2013-12-17  Eric Gallager  <egallager@comcast.net>

	touch more old files

	Change-Id: Idd585fa0166e05dd8af174c0db2e22e96e1c1087

2013-12-17  Eric Gallager  <egallager@comcast.net>

	use AC_USE_SYSTEM_EXTENSIONS macro in src/configure.ac

	Change-Id: I91a767ccfb82218dc32605e6d9246ded0c3cc517

2013-12-17  Eric Gallager  <egallager@comcast.net>

	touch more old files

	Change-Id: I76b365ef5c79a8903c2264d34f55e032e2824638

2013-12-17  Eric Gallager  <egallager@comcast.net>

	more touching/un-obsoleting old files

	Change-Id: Ia569a5b14f94df33a3560e5ce1ccbf5267cc0c1d

2013-12-17  Eric Gallager  <egallager@comcast.net>

	convert tcl/mac/MoreFiles\ 1.4.9/MoreFilesLib.make into a
	Unix Makefile

	Change-Id: I7dcefaefec8f31ea546701b4997aedf426b5cb6d

2013-12-17  Eric Gallager  <egallager@comcast.net>

	tcl.m4 comparisons

	Change-Id: Ib1e6fdceb6745ce2544cfb4adbfb9091bb0399ef

2013-12-17  Eric Gallager  <egallager@comcast.net>

	touch old files in MoreFiles

	Change-Id: Ibeb6a345fe7961c54044bd9b2139f73473ac971f

2013-12-17  Eric Gallager  <egallager@comcast.net>


src/autogen/agen5/autogen.c

/**
 *  @file autogen.c
 *
 *  Time-stamp:        "2012-01-07 10:40:56 bkorb"
 *
 *  This is the main routine for autogen.
 *
 *  This file is part of AutoGen.
 *  Copyright (c) 1992-2012 Bruce Korb - all rights reserved
 *
 * AutoGen is free software: you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * AutoGen is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
#ifdef HAVE_SYS_RESOURCE_H
#include <sys/resource.h>
#endif

typedef void (void_main_proc_t)(int, char **);

typedef enum {
    EXIT_PCLOSE_WAIT,
    EXIT_PCLOSE_NOWAIT
} wait_for_pclose_enum_t;

#define _State_(n)  #n,
static char const * const state_names[] = { STATE_TABLE };
#undef _State_

static sigjmp_buf  abendJumpEnv;
static int         abendJumpSignal = 0;

typedef void (sighandler_proc_t)(int sig);
static sighandler_proc_t ignore_signal, catch_sig_and_bail;

/* = = = START-STATIC-FORWARD = = = */
static void
inner_main(void * closure, int argc, char ** argv);

static void
exit_cleanup(wait_for_pclose_enum_t cl_wait);

static void
cleanup_and_abort(int sig);

static void
catch_sig_and_bail(int sig);

static void
ignore_signal(int sig);

static void
done_check(void);

static void
setup_signals(sighandler_proc_t * chldHandler,
              sighandler_proc_t * abrtHandler,
              sighandler_proc_t * dfltHandler);
/* = = = END-STATIC-FORWARD = = = */

/**
 * main routine under Guile guidance
 * @param  closure Guile closure parameter.  Not used.
 * @param  argc    argument count
 * @param  argv    argument vector
 */
static void
inner_main(void * closure, int argc, char ** argv)
{
    atexit(done_check);
    initialize(argc, argv);

    procState = PROC_STATE_LOAD_DEFS;
    readDefines();

    /*
     *  Activate the AutoGen specific Scheme functions.
     *  Then load, process and unload the main template
     */
    procState = PROC_STATE_LOAD_TPL;

    {
        tTemplate* pTF = loadTemplate(pzTemplFileName, NULL);

        procState = PROC_STATE_EMITTING;
        processTemplate(pTF);

        procState = PROC_STATE_CLEANUP;
        cleanup(pTF);
    }

    procState = PROC_STATE_DONE;
    setup_signals(SIG_DFL, SIG_IGN, SIG_DFL);
    exit_code = AUTOGEN_EXIT_SUCCESS;
    done_check();
    /* NOTREACHED */
}

/**
 * main() called from _start()
 * @param  argc    argument count
 * @param  argv    argument vector
 * @return nothing -- Guile never returns, but calls exit(2).
 */
int
main(int argc, char ** argv)
{
    /*
     *  IF we've been kicked with a signal,
     *  THEN abort, passing the signal that whacked us.
     */
    if (sigsetjmp(abendJumpEnv, 0) != 0)
        cleanup_and_abort(abendJumpSignal);

    setup_signals(ignore_signal, SIG_DFL, catch_sig_and_bail);
    optionSaveState(&autogenOptions);
    pfTrace = stderr;

    /*
     *  as of 2.0.2, Guile will fiddle with strings all on its own accord.
     *  Coerce the environment into being POSIX ASCII strings so it keeps
     *  its bloody stinking nose out of our data.
     */
    putenv((char*)(void*)LC_ALL_IS_C);

    /*
     *  If GUILE_WARN_DEPRECATED has not been defined, then likely we are
     *  not in a development environment and likely we don't want to give
     *  our users any angst.
     */
    if (getenv(GUILE_WARN_DEP_STR) == NULL)
        putenv((char*)(void*)GUILE_WARN_NO_ENV);

    AG_SCM_BOOT_GUILE(argc, argv, inner_main);

    /* NOT REACHED */
    return EXIT_FAILURE;
}

/**
 *  This code must run regardless of which exit path is taken
 */
static void
exit_cleanup(wait_for_pclose_enum_t cl_wait)
{
    /*
     *  There are contexts wherein this function can get run twice.
     */
    {
        static int exit_cleanup_done = 0;
        if (exit_cleanup_done) {
            if (OPT_VALUE_TRACE > TRACE_NOTHING)
                fprintf(pfTrace, EXIT_CLEANUP_MULLIGAN);
            return;
        }

        exit_cleanup_done = 1;
    }

#ifdef SHELL_ENABLED
    SCM_EVAL_CONST(EXIT_CLEANUP_STR);
#endif

    close_server_shell();

    if (OPT_VALUE_TRACE > TRACE_NOTHING)
        fprintf(pfTrace, EXIT_CLEANUP_DONE_FMT,
                (cl_wait == EXIT_PCLOSE_WAIT)
                ? EXIT_CLEANUP_WAITED : EXIT_CLEANUP_NOWAIT,
                (unsigned int)getpid());

    do  {
        if (pfTrace == stderr)
            break;

        if (! trace_is_to_pipe) {
            fclose(pfTrace);
            break;
        }


        fflush(pfTrace);
        pclose(pfTrace);
        if (cl_wait == EXIT_PCLOSE_WAIT) {
            int status;
            while (wait(&status) > 0)  ;
        }
    } while (0);

    fflush(stdout);
    fflush(stderr);
}

/**
 *  A signal was caught, siglongjmp called and main() has called this.
 *  We do not deallocate stuff so it can be found in the core dump.
 */
static void
cleanup_and_abort(int sig)
{
    if (*pzOopsPrefix != NUL) {
        fputs(pzOopsPrefix, stderr);
        pzOopsPrefix = zNil;
    }

    fprintf(stderr, AG_SIG_ABORT_FMT, sig, strsignal(sig),
            ((unsigned)procState <= PROC_STATE_DONE)
            ? state_names[ procState ] : BOGUS_TAG);

    if (procState == PROC_STATE_ABORTING) {
        exit_cleanup(EXIT_PCLOSE_NOWAIT);
        _exit(sig + 128);
    }

    procState = PROC_STATE_ABORTING;
    setup_signals(SIG_DFL, SIG_DFL, SIG_DFL);

    /*
     *  IF there is a current template, then we should report where we are
     *  so that the template writer knows where to look for their problem.
     */
    if (pCurTemplate != NULL) {
        int line;
        teFuncType fnCd;
        char const * pzFn;
        char const * pzFl;

        if (pCurMacro == NULL) {
            pzFn = PSEUDO_MACRO_NAME_STR;
            line = 0;
            fnCd = -1;

        } else {
            teFuncType f =
                (pCurMacro->funcCode > FUNC_CT)
                    ? FTYP_SELECT : pCurMacro->funcCode;
            pzFn = apzFuncNames[ f ];
            line = pCurMacro->lineNo;
            fnCd = pCurMacro->funcCode;
        }
        pzFl = pCurTemplate->pzTplFile;
        if (pzFl == NULL)
            pzFl = NULL_FILE_NAME_STR;
        fprintf(stderr, AG_ABORT_LOC_FMT, pzFl, line, pzFn, fnCd);
    }

#ifdef HAVE_SYS_RESOURCE_H
    /*
     *  If `--core' has been specified and if we have get/set rlimit,
     *  then try to set the core limit to the "hard" maximum before aborting.
     */
    if (HAVE_OPT(CORE)) {
        struct rlimit rlim;
        getrlimit(RLIMIT_CORE, &rlim);
        rlim.rlim_cur = rlim.rlim_max;
        setrlimit(RLIMIT_CORE, &rlim);
    }
#endif

    exit_cleanup(EXIT_PCLOSE_NOWAIT);
    abort();
}


/**
 *  catch_sig_and_bail catches signals we abend on.  The "siglongjmp"
 *  goes back to the real "main()" procedure and it will call
 *  "cleanup_and_abort()", above.
 */
static void
catch_sig_and_bail(int sig)
{
    switch (procState) {
    case PROC_STATE_ABORTING:
        exit_code = AUTOGEN_EXIT_SIGNAL;

    case PROC_STATE_DONE:
        break;

    default:
        abendJumpSignal = sig;
        exit_code = AUTOGEN_EXIT_SIGNAL;
        siglongjmp(abendJumpEnv, sig);
    }
}


/**
 *  ignore_signal is the handler for SIGCHLD.  If we set it to default,
 *  it will kill us.  If we set it to ignore, it will be inherited.
 *  Therefore, always in all programs set it to call a procedure.
 *  The "wait(3)" call will do magical things, but will not override SIGIGN.
 */
static void
ignore_signal(int sig)
{
    return;
}


/**
 *  This is _always_ called for exit processing.
 *  This only returns if "exit(3C)" is called during option processing.
 *  We have no way of determining the correct exit code.
 *  (Requested version or help exits EXIT_SUCCESS.  Option failures
 *  are EXIT_FAILURE.  We cannot determine here.)
 */
static void
done_check(void)
{
    /*
     *  There are contexts wherein this function can get called twice.
     */
    {
        static int done_check_done = 0;
        if (done_check_done) {
            if (OPT_VALUE_TRACE > TRACE_NOTHING)
                fprintf(pfTrace, DONE_CHECK_REDONE);
            return;
        }
        done_check_done = 1;
    }

    switch (procState) {
    case PROC_STATE_EMITTING:
    case PROC_STATE_INCLUDING:
        /*
         *  A library (viz., Guile) procedure has called exit(3C).
         *  The AutoGen abort paths all set procState to PROC_STATE_ABORTING.
         */
        if (*pzOopsPrefix != NUL) {
            /*
             *  Emit the CGI page header for an error message.  We will rewind
             *  stderr and write the contents to stdout momentarily.
             */
            fputs(pzOopsPrefix, stdout);
            pzOopsPrefix = zNil;
        }

        if (OPT_VALUE_TRACE > TRACE_NOTHING)
            scm_backtrace();

        fprintf(stderr, SCHEME_EVAL_ERR_FMT, pCurTemplate->pzTplFile,
                pCurMacro->lineNo);

        /*
         *  We got here because someone called exit early.
         */
        do  {
#ifndef DEBUG_ENABLED
            out_close(AG_FALSE);
#else
            out_close(AG_TRUE);
#endif
        } while (pCurFp->pPrev != NULL);
        exit_code = AUTOGEN_EXIT_BAD_TEMPLATE;
        break; /* continue failure exit */

    case PROC_STATE_LOAD_DEFS:
        exit_code = AUTOGEN_EXIT_BAD_DEFINITIONS;
        /* FALLTHROUGH */

    default:
        fprintf(stderr, AG_ABEND_STATE_FMT, state_names[procState]);
        goto autogen_aborts;

    case PROC_STATE_ABORTING:
        exit_code = AUTOGEN_EXIT_BAD_TEMPLATE;

    autogen_aborts:
        if (*pzOopsPrefix != NUL) {
            /*
             *  Emit the CGI page header for an error message.  We will rewind
             *  stderr and write the contents to stdout momentarily.
             */
            fputs(pzOopsPrefix, stdout);
            pzOopsPrefix = zNil;
        }
        break; /* continue failure exit */

    case PROC_STATE_OPTIONS:
        /* Exiting in option processing state is verbose enough */
        break;

    case PROC_STATE_DONE:
        break; /* continue normal exit */
    }

    if (pzLastScheme != NULL)
        fprintf(stderr, GUILE_CMD_FAIL_FMT, pzLastScheme);

    /*
     *  IF we diverted stderr, then now is the time to copy the text to stdout.
     *  This is done for CGI mode wherein we produce an error page in case of
     *  an error, but otherwise discard stderr.
     */
    if (cgi_stderr != NULL) {
        do {
            long pos = ftell(stderr);
            char* pz;

            /*
             *  Don't bother with the overhead if there is no work to do.
             */
            if (pos <= 0)
                break;
            pz = AGALOC(pos, "stderr redir");
            rewind(stderr);
            fread( pz, (size_t)1, (size_t)pos, stderr);
            fwrite(pz, (size_t)1, (size_t)pos, stdout);
            AGFREE(pz);
        } while (0);

        unlink(cgi_stderr);
        AGFREE(cgi_stderr);
        cgi_stderr = NULL;
    }

    ag_scribble_deinit();

    if (OPT_VALUE_TRACE > TRACE_NOTHING)
        fprintf(pfTrace, DONE_CHECK_DONE);

    exit_cleanup(EXIT_PCLOSE_WAIT);

    /*
     *  When processing options, return so that the option processing exit code
     *  is used.  Otherwise, terminate execution now with the decided upon
     *  exit code.  (Always EXIT_FAILURE, unless this was called at the end
     *  of inner_main().)
     */
    if (procState != PROC_STATE_OPTIONS)
        _exit(exit_code);
}


LOCAL void
ag_abend_at(char const * pzMsg
#ifdef DEBUG_ENABLED
            , char const * pzFile, int line
#endif
    )
{
    if (*pzOopsPrefix != NUL) {
        fputs(pzOopsPrefix, stderr);
        pzOopsPrefix = zNil;
    }

#ifdef DEBUG_ENABLED
    fprintf(stderr, "Giving up in %s line %d\n", pzFile, line);
#endif

    if ((procState >= PROC_STATE_LIB_LOAD) && (pCurTemplate != NULL)) {
        int l_no = (pCurMacro == NULL) ? -1 : pCurMacro->lineNo;
        fprintf(stderr, ERROR_IN_TPL_FMT, pCurTemplate->pzTplFile, l_no);
    }
    fputs(pzMsg, stderr);
    pzMsg += strlen(pzMsg);
    if (pzMsg[-1] != NL)
        fputc(NL, stderr);

    {
        teProcState oldState = procState;
        procState = PROC_STATE_ABORTING;

        switch (oldState) {
        case PROC_STATE_EMITTING:
        case PROC_STATE_INCLUDING:
        case PROC_STATE_CLEANUP:
            longjmp(fileAbort, FAILURE);
            /* NOTREACHED */
        default:
            exit(EXIT_FAILURE);
            /* NOTREACHED */
        }
    }
}


static void
setup_signals(sighandler_proc_t * chldHandler,
              sighandler_proc_t * abrtHandler,
              sighandler_proc_t * dfltHandler)
{
    struct sigaction  sa;
    int    sigNo  = 1;
#ifdef SIGRTMIN
    const int maxSig = SIGRTMIN-1;
#else
    const int maxSig = NSIG;
#endif

    sa.sa_flags   = 0;
    sigemptyset(&sa.sa_mask);

    do  {
        switch (sigNo) {
            /*
             *  Signal handling for SIGCHLD needs to be ignored.  Do *NOT* use
             *  SIG_IGN to do this.  That gets inherited by programs that need
             *  to be able to use wait(2) calls.  At the same time, we don't
             *  want our children to become zombies.  We may run out of zombie
             *  space.  Therefore, set the handler to an empty procedure.
             *  POSIX oversight.  Allowed to be fixed for next POSIX rev, tho
             *  it is "optional" to reset SIGCHLD on exec(2).
             */
#ifndef SIGCHLD
#  define SIGCHLD SIGCLD
#endif
        case SIGCHLD:
            sa.sa_handler = chldHandler;
            break;

        case SIGABRT:
            sa.sa_handler = abrtHandler;
            break;

            /*
             *  Signals we must leave alone.
             */
        case SIGKILL:
        case SIGSTOP:

            /*
             *  Signals we choose to leave alone.
             */
#ifdef SIGTSTP
        case SIGTSTP:
#endif
            continue;

#if defined(DEBUG_ENABLED)
        case SIGBUS:
        case SIGSEGV:
            /*
             *  While debugging AutoGen, we want seg faults to happen and
             *  trigger core dumps.  Make sure this happens.
             */
            sa.sa_handler = SIG_DFL;
            break;
#endif

            /*
             *  Signals to ignore with SIG_IGN.
             */
        case 0: /* cannot happen, but the following might not be defined */
#ifdef SIGWINCH
        case SIGWINCH:
#endif
#ifdef SIGTTIN
        case SIGTTIN:  /* tty input  */
#endif
#ifdef SIGTTOU
        case SIGTTOU:  /* tty output */
#endif
            sa.sa_handler = SIG_IGN;
            break;

#ifdef DAEMON_ENABLED
# error DAEMON-ization of AutoGen is not ready for prime time
  Choke Me.
        case SIGHUP:
            if (HAVE_OPT(DAEMON)) {
                sa.sa_handler = handleSighup;
                break;
            }
            /* FALLTHROUGH */
#endif

        default:
            sa.sa_handler = dfltHandler;
        }
        sigaction(sigNo,  &sa, NULL);
    } while (++sigNo < maxSig);
}

#ifndef HAVE_STRFTIME
#  include <compat/strftime.c>
#endif

#ifndef HAVE_STRSIGNAL
#  include <compat/strsignal.c>
#endif

LOCAL void *
ao_malloc (size_t sz)
{
    void * res = malloc(sz);
    if (res == NULL) {
        fprintf(stderr, MALLOC_FAIL_FMT, sz);
        exit(EXIT_FAILURE);
    }
    return res;
}


LOCAL void *
ao_realloc (void *p, size_t sz)
{
    void * res = (p == NULL) ? malloc(sz) : realloc(p, sz);
    if (res == NULL) {
        fprintf(stderr, REALLOC_FAIL_FMT, sz, p);
        exit(EXIT_FAILURE);
    }
    return res;
}

LOCAL char *
ao_strdup (char const * str)
{
    char * res = strdup(str);
    if (res == NULL) {
        fprintf(stderr, STRDUP_FAIL_FMT, (int)strlen(str));
        exit(EXIT_FAILURE);
    }
    return res;
}

#ifdef __GNUC__
    void ignore_vars(void);
    void ignore_vars(void) {
        (void)option_load_mode, (void)program_pkgdatadir;
    }
#endif
/*
 * Local Variables:
 * mode: C
 * c-file-style: "stroustrup"
 * indent-tabs-mode: nil
 * End:
 * end of agen5/autogen.c */
